servlet/src/test/java/io/grpc/servlet/TomcatInteropTest.java

```diff
@@ -87,7 +87,9 @@ protected void startServer(ServerBuilder<?> builer) {
             new GrpcServlet(((ServletServerBuilder) builer).buildServletAdapter()))
         .setAsyncSupported(true);
     ctx.addServletMappingDecoded("/*", "TomcatInteropTest");
-    server.getConnector().addUpgradeProtocol(new Http2Protocol());
+    Http2Protocol http2Protocol = new Http2Protocol();
+    http2Protocol.setOverheadCountFactor(0);
```

Comment:
I found this to be necessary, for the same reason as the jetty rate control below - otherwise, the "busy" grpc streams tripped some heuristic where the server thought it was being attacked, and it responded with an error rather than complete the test method. I'll add some documentation here and to the patch.

Reply:
https://stackoverflow.com/a/61454503/860630

> The overhead protection was put in place in response to a collection of CVE's reported against HTTP/2 in the middle of 2019. While Tomcat wasn't directly affected (the malicious input didn't trigger excessive load) we did take steps to block input that matched the malicious profile.
...
> The overhead count starts at -10. For every DATA frame received it is reduced by 1. For every SETTINGS, PRIORITY and PING frame it is increased by overheadCountFactor.If the overhead count goes above 0, the connection is closed.
> In addition, if the average size of a received non-final DATA frame and the previously received DATA frame (on that same stream) is less than overheadDataThreshold then the overhead count is increased by overheadDataThreshold/(average size of current and previous DATA frames). In this way, the smaller the DATA frame, the greater the increase in the overhead. A small number of small non-final DATA frames should be enough to trigger connection closure.

https://tomcat.apache.org/tomcat-9.0-doc/config/http2.html has more information about the various "overhead*" configuration keys and their roles, though `overheadDataThreshold`'s documentation says that _non-final_ data frames increase the count rather than reduce it. This appear to be reflected in code, see tomcat's Http2UpgradeHandler.startRequestBodyFrame:
https://github.com/apache/tomcat/blob/9.0.x/java/org/apache/coyote/http2/Http2UpgradeHandler.java#L1510-L1531
```
        // DATA frames reduce the overhead count ...
        reduceOverheadCount(FrameType.DATA);

        // .. but lots of small payloads are inefficient so that will increase
        // the overhead count unless it is the final DATA frame where small
        // payloads are expected.

        // See also https://bz.apache.org/bugzilla/show_bug.cgi?id=63690
        // The buffering behaviour of some clients means that small data frames
        // are much more frequent (roughly 1 in 20) than expected. Use an
        // average over two frames to avoid false positives.
        if (!endOfStream) {
          //...
```

--

Assuming that we didn't want to fine tune these settings for the sake of tests (which are deliberately pathological in some ways, to confirm the behavior of the implementation). Downstream users would need to correctly configure their servlet container to account for their expected usage vs concern over this class of attack.